### PR TITLE
RL online doc

### PR DESCRIPTION
**Pull request by lihuoran**
# Description

## Linked issue(s)/Pull request(s)

- [issue_number](issue_link)

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [ ] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [ ] Simulation toolkit
- [ ] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [ ] Windows
  - [ ] Mac OS
  - [ ] Linux
- Python version:
  - [ ] 3.6
  - [ ] 3.7
- Key information snapshot(s):

## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [ ] Add/update the related comments
- [ ] Add/update the related tests
- [ ] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: docs/source/key_components/rl_toolkit.rst [hunk 1]
```diff
@@ -58,138 +54,115 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
-      def __call__(self, state):
-          """Select an action based on a state."""
-          raise NotImplementedError
+``Policy`` is the most important concept in reinforcement learning. In MARO, the highest level abstraction of a policy
+object is ``AbsPolicy``. It defines the interface ``get_actions()`` which takes a batch of states a inputs and returns
+corresponding actions.
 
+The simplest type of policy is ``RuleBasedPolicy`` which generates actions by pre-defined rules. ``RuleBasedPolicy``
+is mostly used in naive scenarios. However, in most cases, we need to train the policy by interacting with the
+environment, these are the cases we need to use ``RLPolicy``. In MARO's design, a policy cannot train itself. Instead,
```
Comment (ysqyang) [inline docs/source/key_components/rl_toolkit.rst:68]: ... cases where we need to ...

File: docs/source/key_components/rl_toolkit.rst [hunk 2]
```diff
@@ -58,138 +54,115 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
-      def __init__(self, name)
-          super().__init__()
-          self._name = name
-
-      @abstractmethod
-      def __call__(self, state):
-          """Select an action based on a state."""
-          raise NotImplementedError
+``Policy`` is the most important concept in reinforcement learning. In MARO, the highest level abstraction of a policy
+object is ``AbsPolicy``. It defines the interface ``get_actions()`` which takes a batch of states a inputs and returns
```
Comment (ysqyang) [inline docs/source/key_components/rl_toolkit.rst:63]: a inputs -> as input (I believe input should be singular here, but feel free to double check)

File: docs/source/key_components/rl_toolkit.rst [hunk 3]
```diff
@@ -58,138 +54,115 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
+corresponding actions.
 
+The simplest type of policy is ``RuleBasedPolicy`` which generates actions by pre-defined rules. ``RuleBasedPolicy``
+is mostly used in naive scenarios. However, in most cases, we need to train the policy by interacting with the
+environment, these are the cases we need to use ``RLPolicy``. In MARO's design, a policy cannot train itself. Instead,
+polices could only be trained by :ref:`trainer` (we will introduce trainer later in this page). Therefore, in addition
+to ``get_actions``, ``RLPolicy`` also has a set of training-related interfaces, such as ``step()``, ``get_gradients()``
+and ``set_gradients()``. These interfaces will be called by trainers for training. As you may noticed, currently
```
Comment (ysqyang) [inline docs/source/key_components/rl_toolkit.rst:71]: may noticed -> may have noticed

File: docs/source/key_components/rl_toolkit.rst [hunk 4]
```diff
@@ -58,138 +54,115 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
 
+For now, you may have no idea about the ``q_net`` parameter, but don't worry, we will introduce it in the next section.
 
-.. _policy-manager:
+Model
+-----
 
-Policy Manager
---------------
+The above code snippet creates a ``ValueBasedPolicy`` object. Let's pay our attention to the parameter ``q_net``.
```
Comment (ysqyang) [inline docs/source/key_components/rl_toolkit.rst:96]: pay our attention -> pay attention

File: docs/source/key_components/rl_toolkit.rst [hunk 5]
```diff
@@ -58,138 +54,115 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
+``q_net`` accepts a ``DiscreteQNet`` object, and it serves as the core part of a ``ValueBasedPolicy`` object. In
+other words, ``q_net`` defines the model structure of the Q-network in the value-based policy, and further determines
+the policy's behavior. ``DiscreteQNet`` is an abstract class, and ``MyQNet`` is one of the user-defined implementation
```
Comment (ysqyang) [inline docs/source/key_components/rl_toolkit.rst:99]: _one of the user-defined implementation_ -> _one of the user-defined implementation**s**_
or just 
_a user-defined implementation?_

File: docs/source/key_components/rl_toolkit.rst [hunk 6]
```diff
@@ -58,138 +54,115 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
+- ``DiscreteQNet``: For ``ValueBasedPolicy``.
+- ``DiscretePolicyNet``: For ``DiscretePolicyGradient``.
+- ``ContinuousPolicyNet``: For ``ContinuousPolicyGradient``.
 
-Moreover, in ``data-parallel`` mode, there will be an additional ``task_queue`` to perform auto-balance.
-For each training iteration, policies first send requests to ``task_queue``, then ``task_queue`` assigns
-workers according to the idle status. After assigned a set of ``grad_worker`` IDs, policies submit training
-tasks to their own workers. The ``task_queue`` actually processes task requests as a queue.
+Users should choose the proper types of models according to the type of policies, and then implement their own
+models by inherit the abstract ones (just like ``MyQNet``).
```
Comment (ysqyang) [inline docs/source/key_components/rl_toolkit.rst:110]: inherit -> inheriting

File: docs/source/key_components/rl_toolkit.rst [hunk 7]
```diff
@@ -58,138 +54,115 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
 
-.. image:: ../images/rl/policy_manager.svg
-   :target: ../images/rl/policy_manager.svg
-   :alt: PolicyManager
+There are also some other models for training purpose. For example:
```
Comment (ysqyang) [inline docs/source/key_components/rl_toolkit.rst:112]: purpose -> purposes

File: docs/source/key_components/rl_toolkit.rst [hunk 8]
```diff
@@ -58,138 +54,115 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
-and optimization from the algorithmic aspects of the policy that uses them. For example, the actor-critic algorithm
-does not need to concern itself with how the action probabilities and state values are computed. Subclasses of
-``AbsCoreModel`` provided for use with specific RL algorithms include ``DiscreteQNet`` for DQN, ``DiscretePolicyNet``
-for Policy Gradient, ``DiscreteACNet`` for Actor-Critic and ``ContinuousACNet`` for DDPG.
+Trainer
+-------
 
-The code snippet below shows how to create a model for the actor-critic algorithm with a shared bottom stack:
+When introducing policies, we mentioned that policies cannot train themselves. Instead, they have to be trained
+by external trainers. In MARO, a trainer is corresponding to a kind of RL algorithm, such as DQN, actor-critic,
```
Comment (ysqyang) [inline docs/source/key_components/rl_toolkit.rst:126]: _a trainer corresponds to an RL algorithm_ or _a trainer represents an RL algorithm_?

File: docs/source/key_components/rl_toolkit.rst [hunk 9]
```diff
@@ -58,138 +54,115 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
+           reward_discount=.0,
+           grad_iters=10,
+           critic_loss_cls=torch.nn.SmoothL1Loss,
+           min_logp=None,
+           lam=.0
+       )
+   )
+
+In order to indicate which trainer each policy is trained by, in MARO, we require that the name of the policy
+starts with the name of the trainer responsible for training it. For example, policy ``ac_1.policy_1`` is trained
```
Comment (ysqyang) [inline docs/source/key_components/rl_toolkit.rst:169]: starts -> start
https://forum.wordreference.com/threads/requires-that-something-be-read-grammar.3369520/

File: docs/source/key_components/rl_toolkit.rst [hunk 10]
```diff
@@ -58,138 +54,115 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
+by the trainer called ``ac_1``. Violating this provision will make MARO unable to correctly establish the
```
Comment (ysqyang) [inline docs/source/key_components/rl_toolkit.rst:170]: called -> named

File: docs/source/key_components/rl_toolkit.rst [hunk 11]
```diff
@@ -58,138 +54,115 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
-          self.shared_optim.zero_grad()
-          self.actor_optim.zero_grad()
-          self.critic_optim.zero_grad()
-          loss.backward()
-          self.shared_optim.step()
-          self.actor_optim.step()
-          self.critic_optim.step()
-
-To generate stochastic actions given a batch of states, call ``get_action`` on the model instance: 
+Each trainer has a corresponding ``Param`` class that used to manage all related parameters. For example,
```
Comment (ysqyang) [inline docs/source/key_components/rl_toolkit.rst:136]: that used to -> to

File: docs/source/key_components/rl_toolkit.rst [hunk 12]
```diff
@@ -58,138 +59,122 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
-The ``AbsPolicy`` abstraction is the core component of MARO's RL toolkit. A policy is a an agent's mechanism to select
-actions in its interaction with an environment. MARO allows arbitrary agent-to-policy mapping to make policy sharing
-easily configurable. It is also possible to use a mixture of rule-based policies and ``RLPolicy`` instances. The latter
-provides various policy improvement interfaces to support single-threaded and distributed learning.   
-
-
-.. code-block:: python
+``Policy`` is the most important concept in reinforcement learning. In MARO, the highest level abstraction of a policy
+object is ``AbsPolicy``. It defines the interface ``get_actions()`` which takes a batch of states a inputs and returns
+corresponding actions.
```
Comment (Jinyu-W) [inline docs/source/key_components/rl_toolkit.rst:64]: Add description like `(Here the action is defined by the policy itself, could be a number or a vector or some other types. And would be parsed to the Action object in Env Sampler later)`

File: docs/source/key_components/rl_toolkit.rst [hunk 13]
```diff
@@ -58,138 +59,122 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
 
-  class AbsPolicy(ABC):
-      def __init__(self, name)
-          super().__init__()
-          self._name = name
+The simplest type of policy is ``RuleBasedPolicy`` which generates actions by pre-defined rules. ``RuleBasedPolicy``
+is mostly used in naive scenarios. However, in most cases, we need to train the policy by interacting with the
+environment, these are the cases we need to use ``RLPolicy``. In MARO's design, a policy cannot train itself. Instead,
+polices could only be trained by :ref:`trainer` (we will introduce trainer later in this page). Therefore, in addition
+to ``get_actions``, ``RLPolicy`` also has a set of training-related interfaces, such as ``step()``, ``get_gradients()``
```
Comment (Jinyu-W) [inline docs/source/key_components/rl_toolkit.rst:70]: get_actions()

File: docs/source/key_components/rl_toolkit.rst [hunk 14]
```diff
@@ -58,138 +59,122 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
+and ``set_gradients()``. These interfaces will be called by trainers for training. As you may noticed, currently
+we assume policies are built upon deep learning models, so the training-related interfaces are specifically
+designed for gradient decent.
```
Comment (Jinyu-W) [inline docs/source/key_components/rl_toolkit.rst:73]: descent

File: docs/source/key_components/rl_toolkit.rst [hunk 15]
```diff
@@ -58,138 +59,122 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
 
-      @abstractmethod
-      def __call__(self, state):
-          """Select an action based on a state."""
-          raise NotImplementedError
+``RLPolicy`` is further divided into three types:
+- ``ValueBasedPolicy``: For valued-based policies.
```
Comment (Jinyu-W) [inline docs/source/key_components/rl_toolkit.rst:76]: a blank line is required above

File: docs/source/key_components/rl_toolkit.rst [hunk 16]
```diff
@@ -2,51 +2,52 @@ RL Toolkit
...
-Asynchronous Learning
----------------------
 
-The asynchronous mode consists of a policy server and multiple actors in a client-server architecture. Each actor runs
-its own roll-out loop and periodically sends the collected data to the server. The server uses the data to update the
-policies and responds with the latest policy states. As the actors may differ significantly in speed, the policy server
-only uses data generated by policies sufficiently up-to-date, but always sends the latest policy states to every single
-actor. The asynchronous learning feature is still under development and not ready for production use, as it only
-supports a single policy server and hence does not provide fault tolerence.
+.. figure:: ../images/rl/distributed_training.svg
```
[the remark this answers is not in the document]
Comment (Jinyu-W) [reply docs/source/key_components/rl_toolkit.rst:33]: a redundant `...` in the right lower corner

File: docs/source/key_components/rl_toolkit.rst [hunk 17]
```diff
@@ -58,138 +59,122 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
-does not need to concern itself with how the action probabilities and state values are computed. Subclasses of
-``AbsCoreModel`` provided for use with specific RL algorithms include ``DiscreteQNet`` for DQN, ``DiscretePolicyNet``
-for Policy Gradient, ``DiscreteACNet`` for Actor-Critic and ``ContinuousACNet`` for DDPG.
+Trainer
+-------
 
-The code snippet below shows how to create a model for the actor-critic algorithm with a shared bottom stack:
+When introducing policies, we mentioned that policies cannot train themselves. Instead, they have to be trained
+by external trainers. In MARO, a trainer is corresponding to a kind of RL algorithm, such as DQN, actor-critic,
+and so on. Trainers take interaction experiences and store them in a internal memory, and then use the experiences
```
Comment (Jinyu-W) [inline docs/source/key_components/rl_toolkit.rst:127]: in **an** internal memory

File: docs/source/key_components/rl_toolkit.rst [hunk 18]
```diff
@@ -58,138 +59,122 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
+in the memory to train the policies. Like ``RLPolicy``, trainers are also concrete classed, which means they could
```
Comment (Jinyu-W) [inline docs/source/key_components/rl_toolkit.rst:128]: concrete **classes**?

File: docs/source/key_components/rl_toolkit.rst [hunk 19]
```diff
@@ -58,138 +59,122 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
+be used by configuring parameters. Currently, we have 4 trainers in MARO:
```
Comment (Jinyu-W) [inline docs/source/key_components/rl_toolkit.rst:129]: Still feels strange for the naming of `Trainer`. Here,

there are 4 **trainers** in MARO:
- ...: xxx **algorithm** xxx
- ...: xxx **algorithm** xxx

Mark it as a TODO item?

File: docs/source/key_components/rl_toolkit.rst [hunk 20]
```diff
@@ -58,138 +59,122 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
+           lam=.0
+       )
+   )
+
+In order to indicate which trainer each policy is trained by, in MARO, we require that the name of the policy
+starts with the name of the trainer responsible for training it. For example, policy ``ac_1.policy_1`` is trained
+by the trainer called ``ac_1``. Violating this provision will make MARO unable to correctly establish the
+corresponding relationship between policy and trainer.
+
+More details and examples can be found in the code base.
```
Comment (Jinyu-W) [inline docs/source/key_components/rl_toolkit.rst:173]: Add link to the source code page later?

File: docs/source/key_components/rl_toolkit.rst [hunk 21]
```diff
@@ -58,138 +59,122 @@ easily turned into a roll-out worker or an actor for synchronous and asynchronou
...
+       q_net=MyQNet(state_dim=128, action_num=64),
+   )
 
-      def improve(self):
-          pass
 
+For now, you may have no idea about the ``q_net`` parameter, but don't worry, we will introduce it in the next section.
 
-.. _policy-manager:
+Model
```
Comment (Jinyu-W) [inline docs/source/key_components/rl_toolkit.rst:93]: The `Model` section let me feel like: You did say sth., but it seems like you say nothing. 

哈哈哈感觉好像说了什么东西，又好像什么都没说。但我没想到可以怎么改进_(:з」∠)_

File: docs/source/key_components/rl_toolkit.rst (hunk 17 above)
Comment (ysqyang) [reply docs/source/key_components/rl_toolkit.rst:127]: I believe memory is uncountable in this case, so it should be ``in internal memory``. Feel free to double check.

Comment (lihuoran) [reply docs/source/key_components/rl_toolkit.rst:135]: Use `the` anyway.

File: docs/source/key_components/rl_toolkit.rst (hunk 19 above)
Comment (lihuoran) [reply docs/source/key_components/rl_toolkit.rst:137]: we have 4 trainers in MARO => we have 4 trainers (algorithms) in MARO

**Pull request merged**